### PR TITLE
Refactor hugepage out of tt_SiliconDevice

### PR DESCRIPTION
**Pull request by broskoTT**
Related to #142 
Moved hugepage related code to a separate file and to PCIDevice, to further relieve tt_SiliconDevice.

No API changes are imposed with this change.

- Following have been moved to separate compilation unit, which are OS related: get_num_hugepages, get_available_num_host_mem_channels, find_hugepage_dir, open_hugepage_file
- These have been moved to PCIDevice since they are lower level concepts in our ideal hierarchy: init_hugepage
- Added hugepage_mapping struct to hold hugepage intialization related info.
- ioctl and communication with kernel driver can now be completelly removed from tt_SiliconDevice

### CODE REVIEW

File: device/tt_silicon_driver.cpp [hunk 1]
```diff
@@ -1052,20 +949,22 @@ void tt_SiliconDevice::write_buffer(
...
-    if(hugepage_mapping.at(src_device_id).at(channel)) {
-      log_assert(size <= HUGEPAGE_REGION_SIZE, "write_buffer data has larger size {} than destination buffer {}", size, HUGEPAGE_REGION_SIZE);
-      log_debug(LogSiliconDriver, "Using hugepage mapping at address {} offset {} chan {} size {}",
-        hugepage_mapping.at(src_device_id).at(channel),
-        (address & HUGEPAGE_MAP_MASK),
+    log_assert(m_pci_device_map.find(src_device_id) != m_pci_device_map.end(), "write_buffer: Device id is not a MMIO device");
+
+    hugepage_mapping hugepage_map = m_pci_device_map.at(src_device_id)->get_hugepage_mapping(channel);
+    log_assert(hugepage_map.mapping, "write_buffer: Hugepages are not allocated for src_device_id: {} ch: {}."
+                                             " - Ensure sufficient number of Hugepages installed per device (1 per host mem ch, per device)",
```
Comment (pjanevskiTT) [inline device/tt_silicon_driver.cpp:956]: A lot of this log_* lines are not aligned properly when it breaks into multiple lines, but if that is how linter has formatted this then don't change it

File: device/tt_silicon_driver.cpp [hunk 2]
```diff
@@ -1197,216 +1083,29 @@ void tt_SiliconDevice::set_fallback_tlb_ordering_mode(const std::string& fallbac
...
 
         // Device to Host (multiple channels)
-        for (int channel_id = 0; channel_id < m_num_host_mem_channels; channel_id++) {
-            if (hugepage_mapping.at(logical_id).at(channel_id)) {
-                std::uint32_t region_size = HUGEPAGE_REGION_SIZE;
+        for (int channel_id = 0; channel_id < src_pci_device->get_num_host_mem_channels(); channel_id++) {
+            hugepage_mapping hugepage_map = src_pci_device->get_hugepage_mapping(channel_id);
+            if (hugepage_map.mapping) {
+                std::uint32_t region_size = hugepage_map.mapping_size;
                 if (channel_id == 3) region_size = 805306368; // Remove 256MB from full 1GB for channel 3 (iATU limitation)
```
Comment (pjanevskiTT) [inline device/tt_silicon_driver.cpp:1096]: can you add {} around this if, and also put this number `805306368` in a constant variable

File: device/tt_cluster_descriptor_types.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 #pragma once 
 
+#include <functional>
```
[the remark this answers is not in the document]
Comment (broskoTT) [reply device/tt_cluster_descriptor_types.h:9]: std::hash is defined here.
The compilation passes though on one ubuntu version and not on the other, I guess compilers are slightly different.

File: device/tt_silicon_driver.cpp (hunk 1 above)
Comment (broskoTT) [reply device/tt_silicon_driver.cpp:956]: Linter is still not turned on for these files.
I won't fix this manually, we should turn on the linter gradually.

**Review by pjanevskiTT — APPROVED** [review]

**Comment by blozano-tt** [comment]
@broskoTT I think there might have been "API" changes in this PR.

tt_metal build:
```
tt_metal/llrt/tt_cluster.cpp:215:36: error: use of undeclared identifier 'get_num_hugepages'
  215 |     uint32_t total_num_hugepages = get_num_hugepages();
```

**Pull request merged**